instrumentation/github.com/aws/aws-lambda-go/otellambda/wrapLambdaHandler.go

```diff
@@ -0,0 +1,188 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otellambda
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// wrappedHandlerFunction is a struct which only holds an instrumentor and is
+// able to instrument invocations of the user's lambda handler function
+type wrappedHandlerFunction struct {
+	instrumentor instrumentor
+}
+
+func errorHandler(e error) func(context.Context, interface{}) (interface{}, error) {
+	return func(context.Context, interface{}) (interface{}, error) {
+		return nil, e
+	}
+}
+
+// Ensure handler takes 0-2 values, with context
+// as its first value if two arguments exist
+func validateArguments(handler reflect.Type) (bool, error) {
+	handlerTakesContext := false
+	if handler.NumIn() > 2 {
+		return false, fmt.Errorf("handlers may not take more than two arguments, but handler takes %d", handler.NumIn())
+	} else if handler.NumIn() > 0 {
+		contextType := reflect.TypeOf((*context.Context)(nil)).Elem()
+		argumentType := handler.In(0)
+		handlerTakesContext = argumentType.Implements(contextType)
+		if handler.NumIn() > 1 && !handlerTakesContext {
+			return false, fmt.Errorf("handler takes two arguments, but the first is not Context. got %s", argumentType.Kind())
+		}
+	}
+
+	return handlerTakesContext, nil
```

Comment:
Is it intentional that this returns `false` in the event that there are 0 arguments? If so, should we also provide an error message?

Reply:
Yes this is intentional. This return is used to determine whether to pass context to the user's function. The user's function only needs to take context if they plan to instrument downstream operations. Not taking context is a valid choice such as in a case where the Lambda function is the final operation in a trace. Such cases are admittedly unlikely though if they are bothering to instrument lambda.

Reply:
Ok I see, so it's a silent failure by design. That makes sense.